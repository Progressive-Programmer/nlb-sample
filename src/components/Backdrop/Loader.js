import React from 'react';
import PropTypes from 'prop-types';
import Backdrop from '@mui/material/Backdrop';
import CircularProgress from '@mui/material/CircularProgress';

const Loader = (props) => {
  const { isOpen } = props;

  return (
    <Backdrop sx={{ color: '#fff', zIndex: (theme) => theme.zIndex.drawer + 1 }} open={isOpen}>
      <CircularProgress color="inherit" />
    </Backdrop>
  );
};

export default Loader;

Loader.propTypes = {
  isOpen: PropTypes.bool
};

Loader.defaultProps = {
  isOpen: false
};
