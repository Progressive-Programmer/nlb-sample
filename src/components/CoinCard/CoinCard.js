import React, { useState } from "react";
import parse from "html-react-parser";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import Avatar from "@mui/material/Avatar";

import { styled } from "@mui/material/styles";
import Card from "@mui/material/Card";
import CardHeader from "@mui/material/CardHeader";
import CardContent from "@mui/material/CardContent";
import CardActions from "@mui/material/CardActions";
import Collapse from "@mui/material/Collapse";
import IconButton from "@mui/material/IconButton";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

import Grid from "@mui/material/Unstable_Grid2";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableRow from "@mui/material/TableRow";
import Link from "@mui/material/Link";
import CloseIcon from "@mui/icons-material/Close";

const style = {
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  overflow: "scroll",
  minWidth: 600,
  maxHeight: 600,
  bgcolor: "background.paper",
  boxShadow: 24,
  p: 4,
};

const ExpandMore = styled((props) => {
  const { expand, ...other } = props;
  return <IconButton {...other} />;
})(({ theme, expand }) => ({
  transform: !expand ? "rotate(0deg)" : "rotate(180deg)",
  marginLeft: "auto",
  transition: theme.transitions.create("transform", {
    duration: theme.transitions.duration.shortest,
  }),
}));

const CoinCard = (props) => {
  const { isLoading, isOpen, data, setIsModalOpen } = props;
  const [expanded, setExpanded] = useState(false);

  const handleClose = () => {
    setIsModalOpen(false);
    setExpanded(false); //closes description accordion as well when box closes
  };

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  return (
    <div>
      <Modal
        open={isOpen}
        onClose={handleClose}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Card sx={style}>
          <CardHeader // Modal Header
            avatar={<Avatar alt={data?.name} src={data?.image?.thumb} />}
            title={data?.name}
            subheader={data?.symbol}
            action={
              <IconButton aria-label="close" onClick={handleClose}>
                <CloseIcon />
              </IconButton>
            }
          />
          <CardContent>
            <Grid
              container
              rowSpacing={1}
              columnSpacing={{ xs: 1, sm: 2, md: 3 }}
            >
              <Grid xs={6}>
                <Table aria-label="a dense table" size="small">
                  <TableBody>
                    <TableRow key={`${data.name}-1`} sx={{ th: { border: 0 } }}>
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          <Link href={data?.links?.homepage}>Homepage</Link>
                        </Typography>
                      </TableCell>
                    </TableRow>
                    <TableRow
                      key={`${data.name}-2`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Market cap in Euro
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.market_cap?.eur ? data.market_cap.eur : "--"}
                      </TableCell>
                    </TableRow>
                  </TableBody>
                </Table>
              </Grid>
              <Grid xs={6}>
                <Table aria-label="a dense table" size="small">
                  <TableBody>
                    <TableRow
                      key={`${data.name}-3`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Hashing Algorithm
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.hashing_algorithm
                          ? data.hashing_algorithm
                          : "--"}
                      </TableCell>
                    </TableRow>
                    <TableRow
                      key={`${data.name}-4`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Genesis Date
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.genesis_date ? data.genesis_date : "--"}
                      </TableCell>
                    </TableRow>
                  </TableBody>
                </Table>
              </Grid>
            </Grid>
          </CardContent>
          {/* Description Accordion  */}
          <CardActions disableSpacing>
            <ExpandMore
              expand={expanded}
              onClick={handleExpandClick}
              aria-expanded={expanded}
              aria-label="show more"
            >
              <ExpandMoreIcon />
            </ExpandMore>
          </CardActions>
          <Collapse in={expanded} timeout="auto" unmountOnExit>
            <CardContent>
              <Typography paragraph>Description:</Typography>
              <Typography paragraph variant="body2" color="text.secondary">
                {parse(data?.description?.en)}
              </Typography>
            </CardContent>
          </Collapse>
        </Card>
      </Modal>
    </div>
  );
};

export default CoinCard;
