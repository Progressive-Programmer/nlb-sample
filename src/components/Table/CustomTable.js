import React, { useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Avatar from "@mui/material/Avatar";
import ListItemAvatar from "@mui/material/ListItemAvatar";
import Loader from "../Backdrop/Loader";
import CoinCard from "../CoinCard/CoinCard";

const CustomTable = (props) => {
  const { isLoading, tableData, columns, coinData, onPageChange, onRowClick } =
    props;
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [controller, setController] = useState({
    page: 0,
    rowsPerPage: 10,
  });

  useEffect(() => {
    // fetch table data on first render and
    // then on every page change
    getData();
  }, [controller]);

  const getData = async () => {
    // material ui default page count starts with zero
    // and the first page for api is one
    await onPageChange(controller.page + 1, controller.rowsPerPage);
  };

  const handleChangePage = (event, newPage) => {
    setController({
      ...controller,
      page: newPage,
    });
  };

  const handleChangeRowsPerPage = (event) => {
    setController({
      ...controller,
      rowsPerPage: parseInt(event.target.value, 10),
      page: 0,
    });
  };

  const handleClickRow = (row) => {
    onRowClick(row.id);
    setIsModalOpen(true);
  };

  return (
    <Paper sx={{ width: "90%", overflow: "hidden" }}>
      {isLoading ? <Loader isOpen={isLoading} /> : null}
      <TableContainer sx={{ maxHeight: 650 }}>
        <Table stickyHeader aria-label="sticky table" size="small">
          <TableHead>
            <TableRow>
              {columns &&
                columns.map(
                  (
                    column // set custom column names
                  ) => (
                    <TableCell
                      key={column.id}
                      align={column.align}
                      style={{ minWidth: column.minWidth }}
                    >
                      {column.label}
                    </TableCell>
                  )
                )}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData
              ? tableData // maping data to each custom row
                  .map((row, i) => {
                    return (
                      <TableRow
                        hover
                        role="checkbox"
                        tabIndex={-1}
                        key={`${row.code}-${i}`}
                        style={{ cursor: "pointer" }}
                        onClick={() => handleClickRow(row)}
                      >
                        {columns &&
                          columns.map((column) => {
                            const value = row[column.id];
                            if (column.id === "image") {
                              // table cell spearation for image type and text
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  <ListItemAvatar>
                                    <Avatar alt={row.name} src={value} />
                                  </ListItemAvatar>
                                </TableCell>
                              );
                            } else {
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  {column.format && typeof value === "number"
                                    ? column.format(value)
                                    : value}
                                </TableCell>
                              );
                            }
                          })}
                      </TableRow>
                    );
                  })
              : "No Data Available"}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={100}
        rowsPerPage={controller.rowsPerPage}
        page={controller.page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
      {!isLoading && isModalOpen && (
        <CoinCard
          isLoading={isLoading}
          isOpen={isModalOpen}
          data={coinData ? coinData : {}}
          setIsModalOpen={setIsModalOpen}
        />
      )}
    </Paper>
  );
};

export default CustomTable;
