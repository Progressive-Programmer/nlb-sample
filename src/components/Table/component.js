import React, { useCallback, useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Avatar from "@mui/material/Avatar";
import ListItemAvatar from "@mui/material/ListItemAvatar";
import Loader from "../Backdrop/component";

const StickyHeadTable = (props) => {
  const { isLoading, hasError, currentPage, tableData, columns, onPageChange } = props;
  const [controller, setController] = useState({
    page: 0,
    rowsPerPage: 10,
  });

  useEffect(() => {
    getData();
  }, [controller]);

  const getData = async () => {
    try {
      await onPageChange(controller.page+1, controller.rowsPerPage);
    } catch (error) {
      console.log(error);
    }
  };

  const handleChangePage = (event, newPage) => {
    setController({
      ...controller,
      page: newPage,
    });
  };

  const handleChangeRowsPerPage = (event) => {
    setController({
      ...controller,
      rowsPerPage: parseInt(event.target.value, 10),
      page: 0,
    });
  };

  return (
    <Paper sx={{ width: "90%", overflow: "hidden" }}>
      {isLoading ? <Loader isOpen={isLoading} /> : null}
      <TableContainer sx={{ maxHeight: 650 }}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {columns && columns.map(
                (
                  column // set custom column names
                ) => (
                  <TableCell
                    key={column.id}
                    align={column.align}
                    style={{ minWidth: column.minWidth }}
                  >
                    {column.label}
                  </TableCell>
                )
              )}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData && tableData // maping data to each custom row
              .map((row, i) => {
                return (
                  <TableRow
                    hover
                    role="checkbox"
                    tabIndex={-1}
                    key={`${row.code}-${i}`}
                  >
                    {columns && columns.map((column) => {
                      const value = row[column.id];
                      if (column.id === "image") {
                        // table cell spearation for image type and text
                        return (
                          <TableCell key={column.id} align={column.align}>
                            <ListItemAvatar>
                              <Avatar alt={row.name} src={value} />
                            </ListItemAvatar>
                          </TableCell>
                        );
                      } else {
                        return (
                          <TableCell key={column.id} align={column.align}>
                            {column.format && typeof value === "number"
                              ? column.format(value)
                              : value}
                          </TableCell>
                        );
                      }
                    })}
                  </TableRow>
                );
              })}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={100}
        rowsPerPage={controller.rowsPerPage}
        page={controller.page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
    </Paper>
  );
};

export default StickyHeadTable;
