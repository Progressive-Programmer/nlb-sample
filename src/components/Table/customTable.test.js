import React from 'react';
import { shallow } from 'enzyme';
import Table from '@mui/material/Table';
import TableRow from '@mui/material/TableRow';
import TablePagination from '@mui/material/TablePagination';
import CoinCard from '../CoinCard/CoinCard';
import CustomTable from './CustomTable';
import Loader from '../Backdrop/Loader';

const setUp = (props = {}) => {
  const wrapper = shallow(<CustomTable {...props} />);
  return wrapper;
};

describe('CustomTable', () => {
  describe('has props', () => {
    let wrapper;
    let props;
    let state;
    beforeEach(() => {
      props = {
        isLoading: false,
        currentPage: 0,
        tableData: [
          { id: 1, image: 'image', name: 'name', symbol: 'byt', current_price: 1 },
          { id: 2, image: 'image', name: 'name', symbol: 'byt', current_price: 1 },
          { id: 3, image: 'image', name: 'name', symbol: 'byt', current_price: 1 }
        ],
        coinData: {},
        onPageChange: jest.fn(),
        onRowClick: jest.fn(),
        columns: [
          { id: 'image', label: 'Coin', minWidth: 100 },
          { id: 'name', label: 'Name', minWidth: 100 },
          {
            id: 'symbol',
            label: 'Symbol',
            minWidth: 100
          },
          {
            id: 'current_price',
            label: 'Current Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          },
          {
            id: 'high_24h',
            label: 'High 24 Hr Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          },
          {
            id: 'low_24h',
            label: 'Low 24 Hr Price',
            minWidth: 170,
            align: 'right',
            format: (value) => `$${value}`
          }
        ]
      };
      state = {
        controller: {
          page: 0,
          rowsPerPage: 10
        },
        isModalOpen: false,
        setController: jest.fn(),
        setIsModalOpen: jest.fn()
      };
      wrapper = setUp(props, state);
    });
    afterEach(() => {
      jest.resetAllMocks();
    });

    it('should render Table', () => {
      const component = wrapper;
      expect(component.find(Table)).toHaveLength(1);
    });
    it('should render Loader when isLoading true', () => {
      const component = shallow(<CustomTable {...props} isLoading={true} />);
      expect(component.find(Loader)).toHaveLength(1);
    });
    it('should not render Table when TableData is not passed', () => {
      const component = shallow(<CustomTable {...props} tableData={null} />);
      expect(component.find('[role="checkbox"]')).toHaveLength(0);
      expect(component.text('No Data Available')).toBeTruthy();
    });
    it('will open CoinCard when isLoading is false and Row is clicked', () => {
      const onRowClick = props.onRowClick();
      const handleOnRowClick = jest.spyOn(props, 'onRowClick');

      const component = shallow(<CustomTable {...props} onRowClick={onRowClick} />);

      component.find(TableRow).at(1).simulate('click');

      expect(handleOnRowClick).toBeCalled(); // onClick event on the row is called
      expect(component.find(CoinCard)).toHaveLength(1); // coincard component should reder after the api call
    });
    it('will call onPageChange when page is changed', () => {
      const pageNumber = 0;
      const perPage = 10;
      const onPageChange = props.onPageChange(pageNumber + 1, perPage);
      const handleOnPageChange = jest.spyOn(props, 'onPageChange');

      const component = shallow(<CustomTable {...props} onPageChange={onPageChange} />);
      component.find(TablePagination).at(0).simulate('onPageChange');

      expect(handleOnPageChange).toHaveBeenCalledWith(pageNumber + 1, perPage); // onClick event on the row is called
    });
  });
});
