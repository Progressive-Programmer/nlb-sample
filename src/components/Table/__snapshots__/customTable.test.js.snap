// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTable has props should render correctly 1`] = `
<ForwardRef(Paper)
  sx={
    Object {
      "overflow": "hidden",
      "width": "90%",
    }
  }
>
  <ForwardRef(TableContainer)
    sx={
      Object {
        "maxHeight": 650,
      }
    }
  >
    <ForwardRef(Table)
      aria-label="sticky table"
      size="small"
      stickyHeader={true}
    >
      <ForwardRef(TableHead)>
        <ForwardRef(TableRow)>
          <ForwardRef(TableCell)
            key="image"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Coin
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="name"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Name
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="symbol"
            style={
              Object {
                "minWidth": 100,
              }
            }
          >
            Symbol
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="current_price"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Current Price
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="high_24h"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            High 24 Hr Price
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="low_24h"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Low 24 Hr Price
          </ForwardRef(TableCell)>
        </ForwardRef(TableRow)>
      </ForwardRef(TableHead)>
      <ForwardRef(TableBody)>
        <ForwardRef(TableRow)
          hover={true}
          key="undefined-0"
          onClick={[Function]}
          role="checkbox"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          tabIndex={-1}
        >
          <ForwardRef(TableCell)
            key="image"
          >
            <ForwardRef(ListItemAvatar)>
              <ForwardRef(Avatar)
                alt="name"
                src="image"
              />
            </ForwardRef(ListItemAvatar)>
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="name"
          >
            name
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="symbol"
          >
            byt
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="current_price"
          >
            $1
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="high_24h"
          />
          <ForwardRef(TableCell)
            align="right"
            key="low_24h"
          />
        </ForwardRef(TableRow)>
        <ForwardRef(TableRow)
          hover={true}
          key="undefined-1"
          onClick={[Function]}
          role="checkbox"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          tabIndex={-1}
        >
          <ForwardRef(TableCell)
            key="image"
          >
            <ForwardRef(ListItemAvatar)>
              <ForwardRef(Avatar)
                alt="name"
                src="image"
              />
            </ForwardRef(ListItemAvatar)>
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="name"
          >
            name
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="symbol"
          >
            byt
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="current_price"
          >
            $1
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="high_24h"
          />
          <ForwardRef(TableCell)
            align="right"
            key="low_24h"
          />
        </ForwardRef(TableRow)>
        <ForwardRef(TableRow)
          hover={true}
          key="undefined-2"
          onClick={[Function]}
          role="checkbox"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          tabIndex={-1}
        >
          <ForwardRef(TableCell)
            key="image"
          >
            <ForwardRef(ListItemAvatar)>
              <ForwardRef(Avatar)
                alt="name"
                src="image"
              />
            </ForwardRef(ListItemAvatar)>
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="name"
          >
            name
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            key="symbol"
          >
            byt
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="current_price"
          >
            $1
          </ForwardRef(TableCell)>
          <ForwardRef(TableCell)
            align="right"
            key="high_24h"
          />
          <ForwardRef(TableCell)
            align="right"
            key="low_24h"
          />
        </ForwardRef(TableRow)>
      </ForwardRef(TableBody)>
    </ForwardRef(Table)>
  </ForwardRef(TableContainer)>
  <ForwardRef(TablePagination)
    component="div"
    count={100}
    onPageChange={[Function]}
    onRowsPerPageChange={[Function]}
    page={0}
    rowsPerPage={10}
    rowsPerPageOptions={
      Array [
        10,
        25,
        100,
      ]
    }
  />
</ForwardRef(Paper)>
`;
